plugins/ccgx/fu-ccgx-hpi-device.c

```diff
@@ -562,41 +907,159 @@ fu_ccgx_hpi_device_prepare_firmware (FuDevice *device,
+			     "silicon id mismatch, expected 0x%x, got 0x%x",
+			     self->silicon_id, fw_silicon_id);
+		return NULL;
+	}
+	fw_app_type = fu_ccgx_firmware_get_app_type (FU_CCGX_FIRMWARE (firmware));
+	if (fw_app_type != self->fw_app_type) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "app type mismatch, expected 0x%x, got 0x%x",
+			     self->fw_app_type, fw_app_type);
+		return NULL;
+	}
+	fw_mode = fu_ccgx_firmware_get_fw_mode (FU_CCGX_FIRMWARE (firmware));
+	if (fw_mode != fu_ccgx_fw_mode_get_alternate (self->fw_mode)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "FWMode mismatch, expected %s, got %s",
+			     fu_ccgx_fw_mode_to_string (fu_ccgx_fw_mode_get_alternate (self->fw_mode)),
+			     fu_ccgx_fw_mode_to_string (fw_mode));
+		return NULL;
+	}
+	return g_steal_pointer (&firmware);
+}
+
+static gboolean
+fu_ccgx_hpi_get_metadata_offset (FuCcgxHpiDevice *self,
+				 FWMode fw_mode,
+				 guint32 *addr,
+				 guint32 *offset,
+				 GError **error)
+{
+	guint32 addr_max;
+
+	/* sanity check */
+	if (self->flash_row_size == 0x0) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "unset support row size: 0x%x",
+			     self->flash_row_size);
+		return FALSE;
+	}
+
+	/* get the row offset for the flash size */
+	addr_max = self->flash_size / self->flash_row_size;
+	if (self->flash_row_size == 128) {
+		*offset = HPI_META_DATA_OFFSET_ROW_128;
+	} else if (self->flash_row_size == 256) {
+		*offset = HPI_META_DATA_OFFSET_ROW_256;
+	} else {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "unsupported support row size: 0x%x",
+			     self->flash_row_size);
+		return FALSE;
+	}
+
+	/* get the row offset in the flash */
+	switch (fw_mode) {
+	case FW_MODE_FW1:
+		*addr = addr_max - 1;
+		break;
+	case FW_MODE_FW2:
+		*addr = addr_max - 2;
+		break;
+	default:
+		g_set_error_literal (error,
 				     FWUPD_ERROR,
 				     FWUPD_ERROR_NOT_SUPPORTED,
-				     "app type mismatch on image %u, "
-				     "expected 0x%x, got 0x%x",
-				     i, self->fw_app_type, fw_app_type);
-			return NULL;
-		}
+				     "boot recovery not supported");
+		return FALSE;
 	}
+	return TRUE;
+}
 
-	return g_steal_pointer (&firmware);
+/* this will only work after fu_ccgx_hpi_enter_flash_mode() has been used */
+static gboolean
+fu_ccgx_hpi_load_metadata (FuCcgxHpiDevice *self,
+			   FWMode fw_mode,
+			   CCGxMetaData *metadata,
+			   GError **error)
+{
+	guint32 addr = 0x0;
+	guint32 md_offset = 0x0;
+	g_autofree guint8 *buf = NULL;
+
+	/* read flash at correct address */
+	if (!fu_ccgx_hpi_get_metadata_offset (self, fw_mode, &addr, &md_offset, error))
+		return FALSE;
+	buf = g_malloc0 (self->flash_row_size);
+	if (!fu_ccgx_hpi_read_flash (self, addr,
+				     buf, self->flash_row_size,
+				     error)) {
+		g_prefix_error (error, "fw metadata read error: ");
```

Comment:
I think you should use a unique prefix to help with debugging in the future.

Reply:
Fixed.